internal/pkg/configmanager/config.go

```diff
@@ -2,38 +2,138 @@ package configmanager
 
 import (
 	"fmt"
+
+	"gopkg.in/yaml.v3"
+
+	"github.com/devstream-io/devstream/pkg/util/log"
+	"github.com/devstream-io/devstream/pkg/util/mapz"
 )
 
-// Config records rendered config values and is used as a general config in DevStream.
-// Also see the rawConfig struct in rawconfig.go, it represents the original config.
+// Config is a general config in DevStream.
 type Config struct {
-	Tools Tools  `yaml:"tools"`
-	State *State `yaml:"state"`
+	Config              CoreConfig         `yaml:"config"`
+	Vars                map[string]any     `yaml:"vars"`
+	Tools               Tools              `yaml:"tools"`
+	Apps                []app              `yaml:"apps"`
+	PipelineTemplates   []pipelineTemplate `yaml:"pipelineTemplates"`
+	pipelineTemplateMap map[string]string  `yaml:"-"`
+}
+
+func (c *Config) getToolsFromApps() (Tools, error) {
+	var tools Tools
+	for _, a := range c.Apps {
+		appTools, err := c.getToolsWithVarsFromApp(a)
+		if err != nil {
+			return nil, err
+		}
+		tools = append(tools, appTools...)
+	}
+	return tools, nil
+}
+
+func (c *Config) getToolsWithVarsFromApp(a app) (Tools, error) {
+	appStr, err := yaml.Marshal(a)
+	if err != nil {
+		return nil, err
+	}
+
+	// 1. render appStr with Vars
+	appRenderStr, err := renderConfigWithVariables(string(appStr), c.Vars)
+	if err != nil {
+		log.Debugf("configmanager/app %s render globalVars %+v failed", appRenderStr, c.Vars)
+		return nil, fmt.Errorf("app render globalVars failed: %w", err)
+	}
+
+	// 2. unmarshal app config for render pipelineTemplate
+	var rawApp app
+	if err = yaml.Unmarshal(appRenderStr, &rawApp); err != nil {
+		return nil, fmt.Errorf("app parse yaml failed: %w", err)
+	}
+
+	err = c.renderPipelineTemplateMap()
```

Comment:
I think it is better to put this line of code into `getToolsFromApps`. Now this way of writing will cause this function to be called `len (apps)` times.

Reply:
Yeah!